File: sem iv/Theory of Computaion/lab/code/c/7_dfa_c_identifier.cpp
// DFA that recognizes a valid C identifier
#include <iostream>
#include <cctype>
using namespace std;

int current_state = 0;

void state0(char c) {
    if (isalpha((unsigned char)c) || c == '_')
        current_state = 1;
    else
        current_state = -1;
}

void state1(char c) {
    if (isalnum((unsigned char)c) || c == '_')
        current_state = 1;
    else
        current_state = -1;
}

int DFA_for_identifier(string z) {
    current_state = 0;
    int len = (int)z.length();
    for (int i = 0; i < len; i++) {
        if (current_state == 0) state0(z[i]);
        else if (current_state == 1) state1(z[i]);
        else return 0;
    }
    return (current_state == 1) ? 1 : 0;
}

bool isKeyword(string z) {
    string keywords[] = {
        "auto","break","case","char","continue","do","default","const",
        "double","for","else","enum","extern","float","if","goto","int",
        "long","return","signed","static","sizeof","short","struct",
        "switch","typedef","union","void","while","volatile","unsigned"
    };
    for (auto &k : keywords)
        if (z == k) return true;
    return false;
}

int main() {
    string a;
    cout << "Enter string: ";
    cin >> a;

    int result;
    if (isKeyword(a))
        result = -1;
    else
        result = DFA_for_identifier(a);

    if (result == 1)
        cout << a << " : Valid Identifier" << endl;
    else
        cout << a << " : Invalid Identifier" << endl;
    return 0;
}

File: sem iv/Theory of Computaion/lab/code/python/7_dfa_c_identifier.py
# DFA that recognizes a valid C identifier

KEYWORDS = {
    "auto", "break", "case", "char", "continue", "do", "default", "const",
    "double", "for", "else", "enum", "extern", "float", "if", "goto", "int",
    "long", "return", "signed", "static", "sizeof", "short", "struct",
    "switch", "typedef", "union", "void", "while", "volatile", "unsigned"
}

def state0(c):
    return 1 if (c.isalpha() or c == '_') else -1

def state1(c):
    return 1 if (c.isalnum() or c == '_') else -1

def dfa_for_identifier(z):
    current_state = 0
    for c in z:
        if current_state == 0:
            current_state = state0(c)
        elif current_state == 1:
            current_state = state1(c)
        else:
            return 0
    return 1 if current_state == 1 else 0

def is_keyword(z):
    return z in KEYWORDS

def main():
    a = input("Enter string: ")
    if is_keyword(a):
        result = -1
    else:
        result = dfa_for_identifier(a)

    if result == 1:
        print(f"{a} : Valid Identifier")
    else:
        print(f"{a} : Invalid Identifier")

if __name__ == "__main__":
    main()
